torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -820,4 +820,45 @@ std::vector<xla::XlaOp> BuildAmpUpdateScale(const xla::XlaOp& current_scale,
   return results;
 }
 
+std::vector<xla::XlaOp> BuildSgdOptimizerStep(
+    const xla::XlaOp& found_inf, const xla::XlaOp& step,
+    const xla::XlaOp& param, const xla::XlaOp& d_p, const xla::XlaOp& buf,
+    const xla::XlaOp& weight_decay, const xla::XlaOp& momentum,
+    const xla::XlaOp& lr, const xla::XlaOp& dampening, bool use_weight_decay,
+    bool use_momentum, bool use_nesterov) {
+  /* XLA version of the SGD algorithm
+   * https://github.com/pytorch/pytorch/blob/master/torch/optim/_functional.py#L162-L180
+   */
+  auto dtype = XlaHelpers::ShapeOfXlaOp(param).element_type();
+  xla::XlaOp one = xla::One(param.builder(), dtype);
+  xla::XlaOp zero = xla::Zero(param.builder(), dtype);
+
+  xla::XlaOp found_inf_cond = xla::Ne(found_inf, zero);
+  xla::XlaOp is_initialized_cond = xla::Ne(step, zero);
+
+  // weight decay
+  xla::XlaOp d_p_compute = use_weight_decay ? d_p + param * weight_decay : d_p;
+  // update momentum buf
+  xla::XlaOp new_buf = buf;
+  if (use_momentum) {
+    xla::XlaOp buf_compute = xla::Select(
+        is_initialized_cond, buf * momentum + d_p_compute * (one - dampening),
+        d_p_compute);
+    d_p_compute =
+        use_nesterov ? d_p_compute + buf_compute * momentum : buf_compute;
+    new_buf = xla::Select(found_inf_cond, buf, buf_compute);
+  }
+  // update param
+  xla::XlaOp new_param =
+      xla::Select(found_inf_cond, param, param - d_p_compute * lr);
```

Comment:
it might be more efficient to do 
```
param -= d_p_compute * lr * xla::Not(found_inf_cond);
```
XLA should handle boolean to int/double promotion pretty well. You can also save a `not_found_inf_cond` since you need that for below computation too.

Reply:
changed

Reply:
I found this trick is buggy when `d_p_compute` has `nan` and multiply it by 0 does not affect those `nan` elements. I've reverted this optimization and enhanced the test to cover this case.